### PR TITLE
Move HAP to SecretBindings

### PR DESCRIPTION
**Pull request by franpog859**
**Description**

Changes proposed in this pull request:

- make HAP operate on SecretBindings instead of Secrets
- adjust e2e/provisioning integration test
- make Subscription Cleanup Job operate on SecretBindings instead of Secrets

**Keep in mind** that all the Secrets' data in the environments should be migrated to the new SecretBinding logic so KEB works fine! It should be communicated and planned well before the merge of this PR

While testing it with the real solution, please, make sure that KEB is able to maintain secrets in different gardener projects properly. I didn't manage to test it manually. If not, it's probably a cluster role minor issue

**Related issue(s)**

See [the issue](https://github.tools.sap/kyma/backlog/issues/1282)

### CODE REVIEW

**Review by koala7659 — COMMENTED** [review]
The code looks OK.

File: components/subscription-cleanup-job/go.mod [hunk 1]
```diff
@@ -3,30 +3,31 @@ module github.com/kyma-project/control-plane/components/subscription-cleanup-job
...
-	k8s.io/utils v0.0.0-20200324210504-a9aa75ae1b89 // indirect
+	k8s.io/api v0.19.6
+	k8s.io/apimachinery v0.19.6
+	k8s.io/client-go v11.0.1-0.20190409021438-1a26190bd76a+incompatible
 )
 
 replace (
 	github.com/dgrijalva/jwt-go => github.com/dgrijalva/jwt-go v1.0.2
-
+	github.com/kyma-project/control-plane => github.com/franpog859/control-plane v0.0.0-20210308125010-7afa0e3bf7d8
```
[the remark this answers is not in the document]
Comment (franpog859) [reply components/subscription-cleanup-job/go.mod:27]: Ayy ayy! 🚢

File: components/kyma-environment-broker/common/gardener/client.go [hunk 2]
```diff
@@ -38,6 +38,11 @@ func NewGardenerSecretsInterface(gardenerClusterCfg *restclient.Config, gardener
 	return gardenerClusterClient.CoreV1().Secrets(gardenerNamespace), nil
 }
```
[the remark this answers is not in the document]
Comment (franpog859) [reply components/kyma-environment-broker/common/gardener/client.go]: Done!

File: components/kyma-environment-broker/common/hyperscaler/account_provider_test.go [hunk 3]
```diff
@@ -19,93 +18,93 @@ func TestGardenerSharedCredentials_Error(t *testing.T) {
...
 		require.NoError(t, err)
-		assert.Equal(t, secret.Labels["dirty"], "true")
+		assert.Equal(t, secretBinding.Labels["dirty"], "true")
 	})
 
-	t.Run("should not mark secret as dirty if internal", func(t *testing.T) {
+	t.Run("should not mark secret binding as dirty if internal", func(t *testing.T) {
 		//given
-		pool, secretMock := newTestAccountPoolWithSecretInternal()
+		pool, secretBindingMock := newTestAccountPoolWithSecretInternal()
```
Comment (jasiu001) [inline components/kyma-environment-broker/common/hyperscaler/account_provider_test.go:40]: nit: maybe `newTestAccountPoolWithSecretBindingInternal`

File: components/kyma-environment-broker/common/hyperscaler/account_pool.go [hunk 4]
```diff
@@ -28,160 +28,151 @@ type Credentials struct {
...
-		return &secrets.Items[0], nil
+	if secretBindings != nil && len(secretBindings.Items) > 0 {
+		return &secretBindings.Items[0], nil
 	}
-
 	return nil, nil
 }
 
-func credentialsFromSecret(secret *apiv1.Secret, hyperscalerType Type) Credentials {
+func credentialsFromBoundSecret(kubernetesInterface kubernetes.Interface, secretBinding *v1beta1.SecretBinding, hyperscalerType Type) (Credentials, error) {
```
[the remark this answers is not in the document]
Comment (franpog859) [reply components/kyma-environment-broker/common/hyperscaler/account_pool.go]: You're totally right ^^

File: components/kyma-environment-broker/common/hyperscaler/account_pool.go [hunk 5]
```diff
@@ -27,161 +26,136 @@ type Credentials struct {
 }
```
Comment (jasiu001) [inline components/kyma-environment-broker/common/hyperscaler/account_pool.go:26]: please move `Credentials` type to the `account_provider` file

File: components/kyma-environment-broker/common/hyperscaler/account_provider.go [hunk 6]
```diff
@@ -41,13 +46,17 @@ func HyperscalerTypeForPlanID(planID string) (Type, error) {
...
-
 	if p.gardenerPool == nil {
 		return Credentials{},
 			errors.New("failed to get Gardener Credentials. Gardener Account pool is not configured")
 	}
 
-	return p.gardenerPool.Credentials(hyperscalerType, tenantName)
+	secretBinding, err := p.gardenerPool.CredentialsSecretBinding(hyperscalerType, tenantName)
+	if err != nil {
+		return Credentials{}, err
```
Comment (jasiu001) [inline components/kyma-environment-broker/common/hyperscaler/account_provider.go:56]: please always wrap an error:
```go
return ... errors.Wrap(err, "while ...")
```
not only here but everywhere where an error is returned

File: components/kyma-environment-broker/common/hyperscaler/account_pool.go [hunk 7]
```diff
@@ -28,160 +28,151 @@ type Credentials struct {
...
+		return Credentials{}, errors.Wrapf(err, "updating secret binding with tenantName: %s", tenantName)
 	}
 
-	return credentialsFromSecret(updatedSecret, hyperscalerType), nil
+	return credentialsFromBoundSecret(p.kubernetesInterface, updatedSecretBinding, hyperscalerType)
 }
 
-func getK8SSecret(secretsClient corev1.SecretInterface, labelSelector string) (*apiv1.Secret, error) {
-	secrets, err := secretsClient.List(metav1.ListOptions{
+func getSecretBinding(secretBindingsClient gardener_apis.SecretBindingInterface, labelSelector string) (*v1beta1.SecretBinding, error) {
```
[the remark this answers is not in the document]
Comment (franpog859) [reply components/kyma-environment-broker/common/hyperscaler/account_pool.go]: Done!

File: components/kyma-environment-broker/common/hyperscaler/account_provider_test.go (hunk 3 above)
Comment (franpog859) [reply components/kyma-environment-broker/common/hyperscaler/account_provider_test.go]: Done!

File: components/kyma-environment-broker/common/hyperscaler/account_provider.go (hunk 6 above)
Comment (franpog859) [reply components/kyma-environment-broker/common/hyperscaler/account_provider.go]: Ayy ayy, capitan ⚓

File: components/kyma-environment-broker/common/hyperscaler/account_pool.go (hunk 5 above)
Comment (franpog859) [reply components/kyma-environment-broker/common/hyperscaler/account_pool.go]: Done!

**Review by jasiu001 — APPROVED** [review]

**Pull request merged**